### DISCUSSION TITLE
Windows Compile Error

### DISCUSSION THREAD

**Discussion by ernst86913** [directus/directus#7493] [Archive]
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues) that already describes my problem.

### Describe the Bug

Running latest version of Directus from source code Aug 18... on Windows.

a) Performed lerna clean
b) Performed Cache clean
c) Deleted NPM Cache Folder
d) Delete node_modules Folder

**1) Express-Pino-Logger not found**

PS C:\Users\rer\Documents\Coding\directus> npx directus start     
Error: Cannot find module 'express-pino-logger'
Require stack:
- C:\Users\rer\Documents\Coding\directus\api\dist\app.js
- C:\Users\rer\Documents\Coding\directus\api\dist\server.js
- C:\Users\rer\Documents\Coding\directus\api\dist\start.js
- C:\Users\rer\Documents\Coding\directus\api\dist\cli\index.js
- C:\Users\rer\Documents\Coding\directus\api\cli.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:94:18)
    at Object.<anonymous> (C:\Users\rer\Documents\Coding\directus\api\dist\app.js:18:47)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'C:\\Users\\rer\\Documents\\Coding\\directus\\api\\dist\\app.js',
    'C:\\Users\\rer\\Documents\\Coding\\directus\\api\\dist\\server.js',       
    'C:\\Users\\rer\\Documents\\Coding\\directus\\api\\dist\\start.js',        
    'C:\\Users\\rer\\Documents\\Coding\\directus\\api\\dist\\cli\\index.js',   
    'C:\\Users\\rer\\Documents\\Coding\\directus\\api\\cli.js'
  ]
}

2) Installing express-pino-logger using npm

3) Running npx directus start again results in:

PS C:\Users\rer\Documents\Coding\directus> npx directus start                  
11:30:45 ­ƒÜ¿ "DB_CLIENT" Environment Variable is missing.

BUT - I have the .env file in API folder... and it was working in the past... any change regarding the location where the .env file should reside?

Thanks,
Roman

### To Reproduce

See above

### What version of Directus are you using?

v9.0.0-rc.90 + Latest Commits

### What version of Node.js are you using?

node: 'v16.7.0', npm: '7.20.3'

### What database are you using?

Postgres

### What browser are you using?

Chrome

### What operating system are you using?

Windows

### How are you deploying Directus?

locally

**Comment by paescuj** [comment]
Have you performed all the steps from https://docs.directus.io/contributing/running-locally/?

**Comment by ernst86913** [comment]
Yes… and everything worked fine till adding recent code commits… 

LG,
Roman

Von meinem iPhone gesendet

**Comment by paescuj** [comment]
Cannot reproduce this with recent commits. Although I think you are not supposed to run `npx directus start` in a local version of the repository. Did you mean to run `npm run dev -- --scope directus --scope @directus/app`?

**Comment by ernst86913** [comment]
... in the past npx directus start worked for me... I think it actually was guidance from Ben or Rijk to do it that way... 

There seems to be something wrong with SHARP for WINDOWS:

Something went wrong installing the "sharp" module
directus: Cannot find module '../build/Release/sharp-win32-x64.node'

And I saw there were some new commits around "Sharp" recently....

**Comment by rijkvanzanten** [comment]
Can you try installing with `npm install --build-from-source`? That should make sure to build those sort of native bindings fresh 🤔

**Comment by ernst86913** [comment]
First... THANKS A LOT !!!  At least I got it running with it... but.... there are still issues:

1) npx directus start results in the following error (express-pino-logger!):

PS C:\Users\rer\Documents\Coding\directus> npx directus start
Error: Cannot find module 'express-pino-logger'
Require stack:
- C:\Users\rer\Documents\Coding\directus\api\dist\app.js
- C:\Users\rer\Documents\Coding\directus\api\dist\server.js
- C:\Users\rer\Documents\Coding\directus\api\dist\cli\index.js
- C:\Users\rer\Documents\Coding\directus\api\cli.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:94:18)
    at Object.<anonymous> (C:\Users\rer\Documents\Coding\directus\api\dist\app.js:18:47)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    'C:\\Users\\rer\\Documents\\Coding\\directus\\api\\dist\\app.js',
    'C:\\Users\\rer\\Documents\\Coding\\directus\\api\\dist\\server.js',       
    'C:\\Users\\rer\\Documents\\Coding\\directus\\api\\dist\\start.js',        
    'C:\\Users\\rer\\Documents\\Coding\\directus\\api\\dist\\cli\\index.js',   
    'C:\\Users\\rer\\Documents\\Coding\\directus\\api\\cli.js'
  ]
}

2) Also DB Migrations don't work :

npx directus database migrate:latest
20:06:38 ­ƒÜ¿ "DB_CLIENT" Environment Variable is missing.

**Reply by rijkvanzanten** [reply]
> Error: Cannot find module 'express-pino-logger'

How are you installing the npm dependencies? Running `npm install` from the root on node 15 or newer should pull everything in using it's workspaces feature, and tie everything together. A `npm run build` after that should make a dist of everything, and after that, you can use `./api/cli.js start` to start the API in _production_.

For what it's worth, using `npx directus ___` might cause unexpected behavior, as `directus` as a package might not exist as you'd expect under the monorepo

**Reply by rijkvanzanten** [reply]
> npx directus database migrate:latest
> 20:06:38 ­ƒÜ¿ "DB_CLIENT" Environment Variable is missing.

This sounds like it has trouble finding the .env file.

I normally have mine under the `api` folder, and then use `npm run cli bootstrap` to install/migrate the database

**Reply by paescuj** [reply]
If you would like to run `npx directus start` then you need to do it within the "api" folder so it will see the ".env" file.

**Reply by rijkvanzanten** [reply]
Does it recognize the local package as the directus bin executable? I'm wondering if workspaces does that level of linking as well.

**Reply by ernst86913** [reply]
I always follow the instructions from here:  https://docs.directus.io/guides/installation/cli/   ... and word fine till now :-)

I used to run the following commands after fetching new code from GIT in my DEVELOPMENT environment:
- npm install
- npx directus start from directus root which started everything... and took the .env from the API folder

Sometimes I needed to clean lerna, cache... delete node_modules... and then do same again as described above.

Maybe the change to MonoRepo changed a few things there... would be great if those changes in commands would be reflected in the docu to save your precious time...

What I did now (after npm install --build-from-source   which seems to have fixed a lot)
- npm install
- npm run build
- npm run dev
... all from directus root

From Directus API folder:
- npx directus database migrate:latest

Everything seems to be running fine... so - THANKS A LOT FOR YOUR AMAZING SUPPORT!

**Reply by rijkvanzanten** [reply]
The docs you linked are for installing the production version of directus, which could definitely interfere with a local development setup if they're in the same folder. I'd recommend exclusively relying on the commands and steps found in https://docs.directus.io/contributing/running-locally/ for the development setup, which seems to have worked as per your last examples 🙂

**Reply by ernst86913** [reply]
Ah - didn't know that this one is existing :-)    What will npm run cli bootstrap do to existing Directus tables / definition? Overwrite? Or keep and just do migrations?

**Reply by paescuj** [reply]
> Does it recognize the local package as the directus bin executable? I'm wondering if workspaces does that level of linking as well.

It does, AFAIU 👍

**Reply by paescuj** [reply]
> Ah - didn't know that this one is existing :-) What will npm run cli bootstrap do to existing Directus tables / definition? Overwrite? Or keep and just do migrations?

It will not overwrite existing tables, it's just doing migrations (if necessary).
